### DISCUSSION TITLE
KalmanSmootherCV smooth parameters question

### DISCUSSION THREAD

**Discussion by raybellwaves** [anitagraser/movingpandas#203] [Q&A] [answered]
I'm taking a look at https://movingpandas.readthedocs.io/en/master/trajectorysmoother.html and have some usage questions. This may stem from my naivety of how stonesoup works e.g.
https://stonesoup.readthedocs.io/en/v0.1b8/auto_tutorials/01_KalmanFilterTutorial.html#a-nearly-constant-velocity-example
https://stonesoup.readthedocs.io/en/v0.1b8/auto_examples/Smooth_Platform_Coordinate_Transitions.html#a-simple-demo-of-constant-acceleration

I would like a dummy guide of what the parameters of `process_noise_std` and `measurement_noise_std` do. i.e. :
 - if you want to do more smoothing increase the `measurement_noise_std` value???
 - if you are smoothing geo data you should apply a tuple with more smoothing in the longitude direction???

I'm using the geo test data to play with  (https://github.com/anitagraser/movingpandas/blob/master/movingpandas/tests/test_trajectory_smoother.py#L94):

```
from fiona.crs import from_epsg
from movingpandas.tests.test_trajectory import Node, make_traj
from movingpandas.trajectory_smoother import KalmanSmootherCV
import movingpandas as mpd

traj = make_traj(
    [
        Node(),
        Node(1, 0.1, second=1),
        Node(2, 0.2, second=2),
        Node(3, 0, second=3),
        Node(3, 3, second=4),
    ],
    crs=from_epsg(4326),
)
traj.plot()
```
<img width="264" alt="Screen Shot 2022-03-01 at 5 12 29 PM" src="https://user-images.githubusercontent.com/17162724/156257747-984c87b9-f18e-43a5-b149-5ad38434e138.png">

Apply the parameter values in the test:
```
result = KalmanSmootherCV(traj).smooth(
    process_noise_std=0.1, measurement_noise_std=10
)
result.plot()
```
<img width="320" alt="Screen Shot 2022-03-01 at 5 17 38 PM" src="https://user-images.githubusercontent.com/17162724/156258309-ebfe8d0b-2ad0-4be2-9c22-3c21a1726178.png">

If I tinker with parameter space it seems decreasing process and increasing noise (as integers) results in more smoothing but this could depend on the data

```
fig = plt.figure()
ax_array = fig.subplots(2, 2)

for i, process in enumerate([0.5, 1]):
    for j, noise in enumerate([1, 10]):
        r = KalmanSmootherCV(traj).smooth(
            process_noise_std=process, measurement_noise_std=noise
        )
        r.plot(ax=ax_array[i, j])
        ax_array[i, j].set_title(f"process={process}, noise={noise}")
fig.tight_layout()
```
<img width="373" alt="Screen Shot 2022-03-01 at 9 24 30 PM" src="https://user-images.githubusercontent.com/17162724/156282698-b749e511-cdb6-43c0-b57b-cc9cbe2f9548.png">

cc. @sglvladi

**Comment by raybellwaves** [comment]
some more info in the examples page https://github.com/anitagraser/movingpandas-examples/blob/main/1-tutorials/10-smoothing-trajectories.ipynb which could be upstreamed to the core docs

**Comment by anitagraser** [comment]
@sglvladi can you maybe shed some lights on Ray's questions?

**Comment by sglvladi** [comment, marked as answer]
Hi both,

The question relates more to literature regarding Bayesian filtering and smoothing. When running the smoother, we treat the individual points of the input trajectory as noisy measurements of an underlying hidden (smoothed) state which we are trying to infer (i.e. the points of the output trajectory).

To do so, we make some assumptions about the kinematic properties of the object (that is: where do we expect the object to be in the next timestep, given an estimate at the previous timestep) and the process by which the measurements were generated. The former is generally referred to as the "transition (a.k.a. dynamic/motion/process) model", and the later as the "measurement model". These models are generally stochastic and therefore encapsulate a measure of the associated uncertainty.

When it comes to the ``KalmanSmootherCV`` class, the transition model is assumed to be a "nearly Constant Velocity (CV) model". This means that we expect the object to maintain nearly constant velocity between timesteps, with some added process noise. That is:

```
v_{k} = v_{k-1} + w_{k}
x_{k} = x_{k-1} + v_{k}*Δt
```
where, ``x_{k}, v_{k}`` are respectively the position and velocity at time ``k``, ``Δt`` is the elapsed time (in seconds) between times ``k`` and ``k-1``, and ``w_{k}`` is the process noise that is modelled as a zero mean Gaussian distribution, with standard deviation equal to ``process_noise_std``. You can think of ``w_{k}`` as random acceleration increments. Therefore, if our trajectory points are in meters, the unit of measure is ``m/s^2``.

Continuing, the measurement model dictates that measurements are assumed to be subject to random noise:
```
z_{k} = x_{k} + u_{k}
```
where ``z_{k}`` is the generated measurement, ``x_{k}`` is the true observed position, and ``u_{k}`` is the measurement noise, modelled (again) as a zero mean Gaussian distribution, with standard deviation equal to ``measurement_noise_std``. The added noise can be seen as random perturbations of the true position, and therefore, if we make the same assumption as above, the unit of ``measurement_noise_std`` is ``m``.

Revisiting my statement in the first paragraph, and drawing relation to the above notation, the individual points of the input trajectory become the ``z_{k}``'s and the output (smooth) trajectory points are the ``x_{k}``'s.

Now let's think what happens when we vary the ``process_noise_std`` and ``measurement_noide_std``. Firstly, assume we keep ``measurement_noide_std`` constant and we vary ``process_noise_std``. When we increase ``process_noise_std``, we are essentially telling the algorithm that we are less certain about our assumptions about the transition model. Therefore, the algorithm will tend to place the output trajectory points "away" from points that conform to the CV model and "closer" to the input trajectory points (therefore making the output less smooth). If we decrease ``process_noise_std``, we are doing the opposite and therefore the output trajectory will become smoother. If we were to set ``process_noise_std`` to 0, the output will tend to a straight line.

Continuing, lets us assume we keep ``process_noise_std`` constant and we vary ``measurement_noise_std``. When we increase ``measurement_noise_std`` we are telling the algorithm that we expected the input trajectory points to be increasingly noisy, and therefore the algorithm will tend to place the smoothed points "away" from the inputs points and closer to points that conform to the CV model (therefore making the output smoother). Conversely, if we decrease ``measurement_noise_std``, we are saying that we expect the input points to be close the the "true" points, therefore the algorithm will tend to place the output points closer to the inputs. If we were to set ``measurement_noise_std`` to 0, the output will be nearly identical to the input.

Having said the above, we can conclude to an answer to the first question:
* "if you want to do more smoothing increase the ``measurement_noise_std`` value???"
  - Generally, yes. You could equally also decrease ``process_noise_std``.Ultimately, both lowering ``process_noise_std`` and/or increasing ``measurement_noise_std`` will lead to smoother trajectories. The ideal balance of parameter values is up to the user and can vary depending on the input.
 
I am not too sure about what is meant by the second question. Could you rephrase, please? Thanks

P.S. Apologies for the comment equivalent of War & Peace. I have tried to explain it in as much detail as possible, without including too much technical jargon.

**Reply by raybellwaves** [reply]
this is great thanks @sglvladi. Apologies I haven't had chance to digest it properly. I'll look forward to understanding this better
